### PR TITLE
[LLVMCPU] Add conv2d nchw tiling to enable vectorization

### PR DESCRIPTION
**Pull request by raikonenfnu**

### CODE REVIEW

**Comment by raikonenfnu** [comment]
@hanhanW I was thinking of adding a simple e2e conv nchw test to ensure that it's working in case upstream breaks some patterns. What are your thoughts? or do you think upstream tests for decomposition and vectorization it's safe enough? If not, where is a good place to add it, seems kind of awkward to put in any of the directory within tests/e2e 😅

**Comment by dcaballe** [comment]
I'm missing context but perhaps you can add a test to `compiler/src/iree/compiler/Codegen/LLVMCPU/test` and only run the applicable pipeline and match the resulting vector IR. You can take a look at transpose_avx2_lowering.mlir. It's doing something similar.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 1]
```diff
@@ -1240,14 +1240,14 @@ static LogicalResult setRootConfig(
   return success();
 }
 
-/// Sets the lowering configuration for linalg.conv_2d_nhwc_hwcf and
-/// linalg.depthwise_conv_2d_nhwc_hwc operations.
+/// Sets the lowering configuration for linalg.conv_2d_nhwc_hwcf,
+/// linalg.conv_2d_nchw_fchw, and linalg.depthwise_conv_2d_nhwc_hwc operations.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1244]: Let's update the comment to `... for conv ops. See below for supported conv ops`. Then we won't have to update it every time. The snippets are close enough, so I think it's fine.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 2]
```diff
@@ -1299,9 +1299,8 @@ static LogicalResult setConvRootConfig(func::FuncOp entryPointFn,
 static SmallVector<int64_t> getConvWorkgroupSizes(func::FuncOp entryPointFn,
                                                   linalg::LinalgOp op,
                                                   int64_t vectorSize) {
-  bool isSupported =
-      isa<linalg::Conv2DNhwcHwcfOp, linalg::DepthwiseConv2DNhwcHwcOp>(
-          op.getOperation());
+  bool isSupported = isa<linalg::Conv2DNhwcHwcfOp, linalg::Conv2DNchwFchwOp,
+                         linalg::DepthwiseConv2DNhwcHwcOp>(op.getOperation());
   (void)isSupported;
   assert(isSupported && "expected conv with nhwc input and hwcf kernel/filter");
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1305]: update the error message. maybe just say that the conv op is not supported.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 3]
```diff
@@ -1334,6 +1333,16 @@ static LogicalResult setRootConfig(func::FuncOp entryPointFn,
   return setConvRootConfig(entryPointFn, convOp, targetTileSizes, vectorSize);
 }
 
+/// Sets the lowering configuration for linalg.conv_2d_nchw_fchw
+/// operations.
+static LogicalResult setRootConfig(func::FuncOp entryPointFn,
+                                   linalg::Conv2DNchwFchwOp convOp) {
+  int64_t vectorSize =
+      getVectorSize(entryPointFn, convOp.getResult(0).getType());
+  SmallVector<int64_t> targetTileSizes = {1, vectorSize * 2, 1, 8, 8, 1, 1};
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1342]: Are the tiling sizes arbitrary values? Or they are roughly tuned?

This needs to be refactored to use `getConvWorkgroupSizes`. I have a WIP commit and I can fix it there. You don't have to fix it in the PR.

**Review by hanhanW — CHANGES_REQUESTED** [review]
We'll need at least two tests here.

1. Add a test to materialize_x86_64_launch_configuration.mlir like [this](https://github.com/iree-org/iree/blob/e91733064bb2b803a32e2994ff34a59314a93018/compiler/src/iree/compiler/Codegen/LLVMCPU/test/materialize_x86_64_launch_configuration.mlir#L616-L653)
2. Add an e2e test. You can add a test to [xla_ops/convolution.mlir](https://github.com/iree-org/iree/blob/main/tests/e2e/xla_ops/convolution.mlir) or add a new test file to `tests/e2e/tosa_ops/`.

**Comment by raikonenfnu** [comment]
> We'll need at least two tests here.
> 
> 1. Add a test to materialize_x86_64_launch_configuration.mlir like [this](https://github.com/iree-org/iree/blob/e91733064bb2b803a32e2994ff34a59314a93018/compiler/src/iree/compiler/Codegen/LLVMCPU/test/materialize_x86_64_launch_configuration.mlir#L616-L653)
> 2. Add an e2e test. You can add a test to [xla_ops/convolution.mlir](https://github.com/iree-org/iree/blob/main/tests/e2e/xla_ops/convolution.mlir) or add a new test file to `tests/e2e/tosa_ops/`.

Sounds good, for materializing launch config test I can do that.
However for the e2e test, xla/convolution is only for mhlo.convolution which is nhwc no? Looking at tosa ops it is also similar, it seems to be lowering to nhwc as seen [here](https://github.com/llvm/llvm-project/blob/67d0d7ac0acb0665d6a09f61278fbcf51f0114c2/mlir/lib/Conversion/TosaToLinalg/TosaToLinalgNamed.cpp#L162). Can I start a new directory for linalg convs to fill with nchw test?

**Comment by raikonenfnu** [comment]
> I'm missing context but perhaps you can add a test to `compiler/src/iree/compiler/Codegen/LLVMCPU/test` and only run the applicable pipeline and match the resulting vector IR. You can take a look at transpose_avx2_lowering.mlir. It's doing something similar.

Thanks for the suggestion, I was thinking more in the line of e2e test for correctness, just because I think this conv2d nchw vectorization pass will be evolving very quickly through the next few weeks as we tune and tweak the passes to gain more performance and come closer to nhwc. But once we get to a stable point for the vectorization, we can add this :)

**Comment by hanhanW** [comment]
Yes, mhlo.convolution might not support NCHW lowering at this moment... Let's start a tests/linalg directory and put the test to there. (I'm not too worried about numerical issues because they should be caught in integration tests. I assume that it's enabled in SHARK Tank nightly tests.)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 3 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1342]: yeap, these are tuned from the previous KernelDispatch for nhwc before getConvWorkgroupSizes was implemented, but I remapped the tile sizes to work with nchw.

**Comment by raikonenfnu** [comment]
> Yes, mhlo.convolution might not support NCHW lowering at this moment... Let's start a tests/linalg directory and put the test to there. (I'm not too worried about numerical issues because they should be caught in integration tests. I assume that it's enabled in SHARK Tank nightly tests. It's still a good thing to do.)

Sounds good! I was thinking would be also good for quick test when we develop more conv2d vectorizations.

**Comment by raikonenfnu** [comment]
@hanhanW added the test, lmk what you think :)

File: tests/e2e/linalg/BUILD [hunk 4]
```diff
@@ -0,0 +1,75 @@
+# Copyright 2019 The IREE Authors
```
Comment (hanhanW) [inline tests/e2e/linalg/BUILD:1]: 2022

**Review by hanhanW — APPROVED** [review]

File: tests/e2e/linalg/BUILD (hunk 4 above)
Comment (raikonenfnu) [reply tests/e2e/linalg/BUILD]: @hanhanW done! :)

**Pull request merged**